crates/vm/src/arch/execution_mode/metered/memory_ctx.rs

```diff
@@ -17,52 +17,81 @@ impl BitSet {
 
     #[inline(always)]
     pub fn insert(&mut self, index: usize) -> bool {
-        let word_index = index / 64;
-        let bit_index = index % 64;
+        let word_index = index >> 6;
+        let bit_index = index & 63;
         let mask = 1u64 << bit_index;
 
-        let was_set = (self.words[word_index] & mask) != 0;
-        self.words[word_index] |= mask;
+        debug_assert!(word_index < self.words.len(), "BitSet index out of bounds");
+
+        // SAFETY: word_index is derived from a memory address that is bounds-checked
+        //         during memory access. The bitset is sized to accommodate all valid
+        //         memory addresses, so word_index is always within bounds.
+        let word = unsafe { self.words.get_unchecked_mut(word_index) };
+        let was_set = (*word & mask) != 0;
+        *word |= mask;
         !was_set
     }
 
     /// Set all bits within [start, end) to 1, return the number of flipped bits.
     #[inline(always)]
     pub fn insert_range(&mut self, start: usize, end: usize) -> usize {
         debug_assert!(start < end);
+        debug_assert!(end <= self.words.len() * 64, "BitSet range out of bounds");
+
         let mut ret = 0;
-        let start_word_index = start / u64::BITS as usize;
-        let end_word_index = (end - 1) / u64::BITS as usize;
-        let start_bit = start as u32 % u64::BITS;
+        let start_word_index = start >> 6;
+        let end_word_index = (end - 1) >> 6;
+        let start_bit = (start & 63) as u32;
+
         if start_word_index == end_word_index {
-            let end_bit = (end - 1) as u32 % u64::BITS + 1;
+            let end_bit = ((end - 1) & 63) as u32 + 1;
             let mask_bits = end_bit - start_bit;
-            let mask = (u64::MAX >> (u64::BITS - mask_bits)) << start_bit;
-            ret += mask_bits - (self.words[start_word_index] & mask).count_ones();
-            self.words[start_word_index] |= mask;
+            let mask = (u64::MAX >> (64 - mask_bits)) << start_bit;
+            // SAFETY: start_word_index is derived from start which is bounds checked by
+            // debug_assert above
```

Comment:
I don't think "checked by debug_assert" is ever a valid reason for safety

Reply:
but this could be an assumption of the function

Reply:
makes sense. i added a comment saying that checking this is responsibility of the caller